File: internal/utils/answer_normalize.go
package utils

import (
	"strings"
	"unicode"

	"golang.org/x/text/unicode/norm"
)

func isCombiningMark(r rune) bool {
	return unicode.Is(unicode.Mn, r) || unicode.Is(unicode.Mc, r) || unicode.Is(unicode.Me, r)
}

// stripUnicodeMarks applies NFD and removes combining marks so typed answers without diacritics
// match canonical names (Czech, Polish, Vietnamese tones, Greek monotonic, etc.).
func stripUnicodeMarks(s string) string {
	nfd := norm.NFD.String(s)
	var b strings.Builder
	b.Grow(len(nfd))
	for _, r := range nfd {
		if isCombiningMark(r) {
			continue
		}
		b.WriteRune(r)
	}
	return b.String()
}

// latinGreekHomoglyph maps ASCII vowels to Greek lowercase letters that look the same, so mixed-script
// typing (e.g. Greek letters with a Latin "o" by mistake) still matches fully Greek canonical names.
var latinGreekHomoglyph = map[rune]rune{
	'a': '\u03b1', 'e': '\u03b5', 'i': '\u03b9', 'o': '\u03bf', 'u': '\u03c5',
}

func foldLatinGreekHomoglyphs(s string) string {
	var b strings.Builder
	b.Grow(len(s))
	for _, r := range s {
		if mapped, ok := latinGreekHomoglyph[r]; ok {
			b.WriteRune(mapped)
		} else {
			b.WriteRune(r)
		}
	}
	return b.String()
}

// NormalizeAnswerForCompare trims whitespace, lowercases using Unicode rules, strips combining marks
// after NFD (accents / Greek tonos / Czech háček, etc.), and folds Latin vowel lookalikes into Greek
// letters so mixed-script Greek guesses still match.
func NormalizeAnswerForCompare(s string) string {
	s = strings.TrimSpace(strings.ToLower(s))
	s = stripUnicodeMarks(s)
	return foldLatinGreekHomoglyphs(s)
}

File: web/src/lib/utils/answerNormalize.ts
/** Latin vowels that look identical to Greek; fold so mixed-script input matches Greek names. */
const latinGreekHomoglyph = new Map<number, number>([
	[0x0061, 0x03b1],
	[0x0065, 0x03b5],
	[0x0069, 0x03b9],
	[0x006f, 0x03bf],
	[0x0075, 0x03c5],
]);

/** NFD then strip all Unicode marks (Mn/Mc/Me), same idea as Go stripUnicodeMarks. */
function stripUnicodeMarks(s: string): string {
	return s.normalize('NFD').replace(/\p{M}/gu, '');
}

function foldLatinGreekHomoglyphs(s: string): string {
	let out = '';
	for (const ch of s) {
		const c = ch.codePointAt(0)!;
		const mapped = latinGreekHomoglyph.get(c);
		out += mapped !== undefined ? String.fromCodePoint(mapped) : ch;
	}
	return out;
}

/**
 * Trims, lowercases (Unicode), strips accents/diacritics via NFD + mark removal, and folds Latin vowel
 * homoglyphs into Greek for answer matching.
 *
 * @param s - Raw user or canonical answer text
 * @returns Normalized string for comparison
 */
export function normalizeAnswerForCompare(s: string): string {
	return foldLatinGreekHomoglyphs(stripUnicodeMarks(s.trim().toLowerCase()));
}
